File: LeetCodeJS/CourseScheduleII_210.js
// courseScheduleII.js
class CourseScheduleII_210 {
    static UNVISITED = 0;
    static VISITING = 1;
    static VISITED = 2;

    findOrder(numCourses, prerequisites) {
        const order = [];
        const graph = new Map();

        // Build the graph
        prerequisites.forEach(([course, prereq]) => {
            if (!graph.has(course)) {
                graph.set(course, []);
            }
            graph.get(course).push(prereq);
        });

        const states = Array(numCourses).fill(CourseScheduleII_210.UNVISITED);

        const dfs = (course) => {
            if (states[course] === CourseScheduleII_210.VISITING) return false; // Cycle detected
            if (states[course] === CourseScheduleII_210.VISITED) return true;    // Already processed

            states[course] = CourseScheduleII_210.VISITING;
            const neighbors = graph.get(course) || [];

            for (const neighbor of neighbors) {
                if (!dfs(neighbor)) return false;
            }

            states[course] = CourseScheduleII_210.VISITED;
            order.push(course); // Add course to order as it completes
            return true;
        };

        for (let i = 0; i < numCourses; i++) {
            if (states[i] === CourseScheduleII_210.UNVISITED && !dfs(i)) {
                return []; // If a cycle is detected, return an empty array
            }
        }

        return order;
    }
}

// Export the CourseScheduleII class
module.exports = CourseScheduleII_210;

File: LeetCodeCpp/problems/graphs/CourseScheduleII_210.cpp
#include "CourseScheduleII_210.h"

#include <functional>
#include <queue>
#include <vector>

using namespace std;

vector<int> CourseScheduleII_210::findOrderByDFSTraversal(
    int numCourses,
    vector<vector<int>>& prerequisites)
{
    vector<vector<int>> graph(numCourses);
    for (const auto& p : prerequisites) {
        const int course = p[0];
        const int pre = p[1];

        graph[course].push_back(pre);
    }

    enum State { UNVISITED, VISITING, VISITED };
    vector<State> states(numCourses, UNVISITED);
    vector<int> order;

    // Recursive lambda kept here to keep the DFS local to this variant.
    function<bool(int)> dfs = [&](int course) {
        if (states[course] == VISITING) {
            return false;
        }

        if (states[course] == VISITED) {
            return true;
        }

        states[course] = VISITING;

        for (int pre : graph[course]) {
            if (!dfs(pre)) {
                return false;
            }
        }

        states[course] = VISITED;

        // With course -> prerequisite edges, DFS postorder already gives prerequisite-first order.
        order.push_back(course);

        return true;
    };

    for (int i = 0; i < numCourses; ++i) {
        if (!dfs(i)) {
            return {};
        }
    }

    return order;
}

vector<int> CourseScheduleII_210::findOrderByKahnsAlgorithm(
    int numCourses,
    vector<vector<int>>& prerequisites)
{
    vector<vector<int>> graph(numCourses);
    vector<int> inDegree(numCourses, 0);

    for (const auto& p : prerequisites) {
        const int course = p[0];
        const int pre = p[1];

        // Kahn needs prerequisite -> course so inDegree[course] counts remaining prerequisites.
        graph[pre].push_back(course);
        ++inDegree[course];
    }

    queue<int> q;
    for (int i = 0; i < numCourses; ++i) {
        if (inDegree[i] == 0) {
            q.push(i);
        }
    }

    vector<int> order;

    while (!q.empty()) {
        const int course = q.front();
        q.pop();

        order.push_back(course);

        for (int next : graph[course]) {
            if (--inDegree[next] == 0) {
                q.push(next);
            }
        }
    }

    return static_cast<int>(order.size()) == numCourses ? order : vector<int>();
}
